crates/extensions/c8y_mapper_ext/src/service_monitor.rs

```diff
@@ -21,182 +15,99 @@ fn default_status() -> String {
     "unknown".to_string()
 }
 
-fn default_type() -> String {
-    "".to_string()
-}
-
-#[derive(Deserialize, Serialize, Debug)]
-pub struct TopicInfo {
-    pub service_name: String,
-    pub child_id: Option<String>,
-}
-
-impl TopicInfo {
-    fn parse_topic_info(topic: &str) -> Self {
-        let topic_split: Vec<&str> = topic.split('/').collect();
-        let service_name = if topic_split.len() == 4 {
-            topic_split[3]
-        } else {
-            topic_split[2]
-        }
-        .to_string();
-
-        let child_id = if topic_split.len() == 4 {
-            Some(topic_split[2].to_owned())
-        } else {
-            None
-        };
-
-        Self {
-            service_name,
-            child_id,
-        }
+pub fn convert_health_status_message(entity: &EntityMetadata, message: &Message) -> Vec<Message> {
+    // TODO: introduce type to remove entity type guards
+    if entity.r#type != EntityType::Service {
+        return vec![];
     }
-}
 
-pub fn convert_health_status_message(
-    message: &Message,
-    device_name: String,
-    default_service_type: String,
-) -> Vec<Message> {
     let mut mqtt_messages: Vec<Message> = Vec::new();
-    let topic = message.topic.name.to_owned();
-    let topic_info = TopicInfo::parse_topic_info(&topic);
-    let default_health_status = format!("\"type\":{default_service_type},\"status\":\"unknown\"");
 
     // If not Bridge health status
-    if !topic_info.service_name.contains("bridge") {
-        let payload_str = message.payload_str().unwrap_or(&default_health_status);
-
-        let mut health_status =
-            serde_json::from_str(payload_str).unwrap_or_else(|_| HealthStatus {
-                service_type: default_service_type.clone(),
-                status: "unknown".to_string(),
-            });
-
-        if health_status.status.is_empty() {
-            health_status.status = "unknown".into();
-        }
-
-        if health_status.service_type.is_empty() {
-            health_status.service_type = if default_service_type.is_empty() {
-                DEFAULT_SERVICE_TYPE.to_string()
-            } else {
-                default_service_type
-            };
-        }
-
-        let status_message = service_monitor_status_message(
-            &device_name,
-            &topic_info.service_name,
-            &health_status.status,
-            &health_status.service_type,
-            topic_info.child_id,
-        );
-
-        mqtt_messages.push(status_message);
+    if entity.topic_id.as_str().contains("bridge") {
+        return mqtt_messages;
     }
 
-    mqtt_messages
-}
+    let payload_str = message.payload_str().expect("payload is valid utf8");
```

Comment:
That `expect` looks dangerous. We can crash the mapper with a malformed payload, unless we transform it to an `unknown` status as done in the old code.

Reply:
Fixed in b0ef0923cf85647526043f916bf99b356ba95532.